web_src/js/features/dropzone.js

```diff
@@ -1,80 +1,131 @@
+import {createElementFromHTML, createElementFromAttrs, elemGetAttributeNumber} from '../utils/dom.js';
 
 const {csrfToken, i18n} = window.config;
 
-export async function createDropzone(el, opts) {
+async function createDropzone(el, opts) {
   const [{Dropzone}] = await Promise.all([
     import(/* webpackChunkName: "dropzone" */'dropzone'),
     import(/* webpackChunkName: "dropzone" */'dropzone/dist/dropzone.css'),
   ]);
   return new Dropzone(el, opts);
 }
 
-export function initGlobalDropzone() {
-  for (const el of document.querySelectorAll('.dropzone')) {
-    initDropzone(el);
-  }
+function addCopyLink(file) {
+  // Create a "Copy Link" element, to conveniently copy the image or file link as Markdown to the clipboard
+  // The "<a>" element has a hardcoded cursor: pointer because the default is overridden by .dropzone
+  const copyLinkEl = createElementFromHTML(`
+<div class="tw-text-center">
+  <a href="#" class="tw-cursor-pointer">${svg('octicon-copy', 14)} Copy link</a>
+</div>`);
+  copyLinkEl.addEventListener('click', async (e) => {
+    e.preventDefault();
+    let fileMarkdown = `[${file.name}](/attachments/${file.uuid})`;
+    if (file.type?.startsWith('image/')) {
+      fileMarkdown = `!${fileMarkdown}`;
+    } else if (file.type?.startsWith('video/')) {
+      fileMarkdown = `<video src="/attachments/${htmlEscape(file.uuid)}" title="${htmlEscape(file.name)}" controls></video>`;
+    }
+    const success = await clippie(fileMarkdown);
+    showTemporaryTooltip(e.target, success ? i18n.copy_success : i18n.copy_error);
+  });
+  file.previewTemplate.append(copyLinkEl);
 }
 
-export function initDropzone(el) {
-  const $dropzone = $(el);
-  const _promise = createDropzone(el, {
-    url: $dropzone.data('upload-url'),
+/**
+ * @param {HTMLElement} dropzoneEl
+ */
+export async function initDropzone(dropzoneEl) {
+  if (!dropzoneEl) return null;
+
+  const listAttachmentsUrl = dropzoneEl.closest('[data-attachment-url]')?.getAttribute('data-attachment-url');
+  const removeAttachmentUrl = dropzoneEl.getAttribute('data-remove-url');
+  const attachmentBaseLinkUrl = dropzoneEl.getAttribute('data-link-url');
+
+  let disableRemovedfileEvent = false; // when resetting the dropzone (removeAllFiles), disable the "removedfile" event
+  let fileUuidDict = {}; // to record: if a comment has been saved, then the uploaded files won't be deleted from server when clicking the Remove in the dropzone
+  const dzInst = await createDropzone(dropzoneEl, {
+    url: dropzoneEl.getAttribute('data-upload-url'),
     headers: {'X-Csrf-Token': csrfToken},
-    maxFiles: $dropzone.data('max-file'),
-    maxFilesize: $dropzone.data('max-size'),
-    acceptedFiles: (['*/*', ''].includes($dropzone.data('accepts'))) ? null : $dropzone.data('accepts'),
+    maxFiles: elemGetAttributeNumber('data-max-file', null), // match dropzone default value, no limit
+    maxFilesize: elemGetAttributeNumber('data-max-size', 256), // match dropzone default value: 256 MiB
```

Comment:
It doesn't make sense to play with our config here.

It is for "dropzone".

It is our code's responsibility to always fill the correct config value into here.

Reply:
Likely it would be better to not set the value at all if it doesn't have a usable value, e.g. optional spread syntax into the object.

Reply:
Reverted all 15ce3f20e1628f8554c0cb253f062911020d9400

Reply:
Will write to remove the non-existing keys

Reply:
Thanks, seems better and I welcome the removal of this helper. Still I'd prefer optional spread:


````js
{
  ...(condition && {value: 1}
}

Reply:
The last time your "optional spread" introduces bugs and I reviewed them out.

* https://github.com/go-gitea/gitea/pull/25253#pullrequestreview-1480581024


I have been enough for such fragile code tricks, and I do not want to introduce more regressions or make the code more unreadable.

Reply:
Behavior is equivalent to your variant, I don't see an issue and I use this style all the time.

Reply:
Maybe you could try by yourself to see how the bug occurs. I think I have explained in the comment.

Reply:
I don't see how that link is related to this method. But I won't block on this either. I would of course use the optional spread syntax for future code.

Reply:
Would you like to introduce some helper functions like this?


```
const opts = {
 ... objectOptionalNumber('max-size', el.getAttribute('data-size')),
}

function objectOptionalNumber(k, v) {
    return v !== null && v !== undefined ? {k: Number(v)} : {};
}

function objectOptionalString(k, v) {
    return v !== null && v !== undefined ? {k: String(v)} : {};
}

```

Otherwise, it would be:

```
 ... el.getAttribute('data-size') === null ? {'max-size', Number(el.getAttribute('data-size'))} : {},
```

It doesn't seem easy to read.

Reply:
Or:

```
function objectOptional(k, cast, v) {
    return v !== null && v !== undefined ? {k: cast(v)} : {};
}

// then:
 ... objectOptional('max-size', Number, el.getAttribute('data-size')),
```

Reply:
Honestly I'm not a fan of such helpers, it forces uninitated users to look up how it's implemented. It's also the reason why I don't like `queryElem` etc.

```js
function objectOptionalNumber(k, v) {
    return v !== null && v !== undefined ? {k: Number(v)} : {};
}
{...objectOptionalNumber("k", str)}
```

Could roughly be implemented as this:

```js
{...(str && {k: Number(str)})}
```

Falsy values don't spread into a object, so the only difference would be that `0` is not spread but it's okay for the cases in this PR.

Reply:
Then where does `str` come from in this case? Either you write 2 lines, or go into a pretty long line:

```
 ... el.getAttribute('data-size') !== null ? {'max-size': Number(el.getAttribute('data-size'))} : {},
```


Could you show how to correctly apply your "optional spread" in this case?

Reply:
I see no clean way to avoid double `el.getAttribute` in this case, so yes, the optional spread would only work if you have a variable like `str`.

Reply:
Then I guess something like `optionalSpread` wins.

And that's why I do not like using optional spread everywhere. In many cases it doesn't fit, and when the expression becomes complex, it is difficult to read and might introduce bugs.